presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java

```diff
@@ -238,20 +243,26 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transaction, Co
                 if ((tableColumns == null) || (partitionColumns == null)) {
                     throw new PrestoException(HIVE_INVALID_METADATA, format("Table '%s' or partition '%s' has null columns", tableName, partName));
                 }
+                ImmutableMap.Builder<Integer, HiveType> mismatchColumnTypes = ImmutableMap.builder();
                 for (int i = 0; i < min(partitionColumns.size(), tableColumns.size()); i++) {
                     HiveType tableType = tableColumns.get(i).getType();
                     HiveType partitionType = partitionColumns.get(i).getType();
                     if (!tableType.equals(partitionType)) {
-                        throw new PrestoException(HIVE_PARTITION_SCHEMA_MISMATCH, format("" +
-                                        "There is a mismatch between the table and partition schemas. " +
-                                        "The column '%s' in table '%s' is declared as type '%s', " +
-                                        "but partition '%s' declared column '%s' as type '%s'.",
-                                tableColumns.get(i).getName(),
-                                tableName,
-                                tableType,
-                                partName,
-                                partitionColumns.get(i).getName(),
-                                partitionType));
+                        if (coercionPolicy.canCoerce(partitionType, tableType)) {
+                            mismatchColumnTypes.put(i, partitionType);
+                        }
+                        else {
+                            throw new PrestoException(HIVE_PARTITION_SCHEMA_MISMATCH, format("" +
```

Comment:
Also tell exactly which partition has he mismatch. This error is a top user question and thus needs to be as easy to understand as possible.